### DISCUSSION TITLE
[Packaging request] Package in Fedora COPR repo

### DISCUSSION THREAD

**Discussion by MalariaKills** [nwg-piotr/nwg-shell#38] [Porting for other distros]
Hello,

I think this project looks really cool, but there is no way package for it right now on Fedora and managing a github package suite would be cumbersome. Is there any interest/possibility in packaging this for Fedora in a COPR (COPR is an automated build system on fedora similar to the AUR on arch)

**Comment by nwg-piotr** [comment]
Well, I'll be glad to help, or even make some changes to the code, if necessary, but I can't package it myself. I've never ever put my hand on Fedora. Please get in touch with the distro maintainers.

**Comment by kuba3351** [comment]
Hello

I can see someone packaged thw whole nwg-shell in fedora copr: https://copr.fedorainfracloud.org/coprs/tofik/nwg-shell/

So, I thing it will be nice to add statuses of the package build and version in the readme. I can see also nwg-panel is submitted in the official fedora repository, I think it will be nice to get other packages available in the official repo.

**Comment by nwg-piotr** [comment]
If I understand the description well, those are just packages, w/o any install script. Correct me if I'm wrong. Otherwise I suggest to consider adding some install script. 

[example](https://github.com/nwg-piotr/nwg-shell/blob/main/install/arch.sh)

**Comment by kuba3351** [comment]
Yes, there are only packages right now, but I can help with writing an install script. Fedora will introduce a fedora Sway Spin soon (https://fedoraproject.org/wiki/Changes/Fedora_Sway_Spin). It is already available on Fedora Rawhide. I tried to install nwg-shell on top of it, I had to just add the copr repo, install nwg-shell package, execute `nwg-shell-installer -a` and restart sway. It worked very well.

**Reply by nwg-piotr** [reply]
Excellent. The latest install script for Arch (linked above) doesn't do much more, as you may see. The difference between `nwg-shell-installer -a` and `nwg-shell-installer -w` is that the latter won't display any warning, nor ask for confirmation. Feel free to submit "fedora.sh" to the same repo directory.

Also, some link to appropriate README would be of use, to mention it somewhere in the shell docs.

**Comment by nwg-piotr** [comment]
[Guides added to the Wiki](https://github.com/nwg-piotr/nwg-shell/wiki#fedora) and the install script provided by @kuba3351 were enough for me, a seasoned Arch user, to install Fedora for the first time in my life, together with the sway session and nwg-shell. Many thanks to @kuba3351 and all the packagers!

There exist some minor issues, and they have been addressed [in the PR comments](https://github.com/nwg-piotr/nwg-shell/pull/68#issuecomment-1416888204). Let's move further discussion to this thread, to keep all the info in one place.

**Comment by nwg-piotr** [comment]
I'll try to sum up all the minor issues I found so far.

- volume slider & device menu don't work (missing `pamixer` package);
- `azote` icon missing from nwg-shell-config (probably due to lack of `/usr/share/pixmaps/azote.svg`);
- XDG User Directories not created during installation (required by nwg-drawer and nwg-menu, as they both search files inside them).

Also: the current nwg-shell-config version (0.4.14) uses the [update indicator](https://github.com/nwg-piotr/nwg-shell-config/blob/master/nwg_shell_config/update_indicator.py) script, that adds a tray update notifier. It would be nice to get it to work on Fedora.

**Comment by Peppe61** [comment]
One small addition from me, for the Sericea version:

- changing the default font in GTK Settings (from Cantarell to any other font) doesn't seem to 'stick'; still Cantarell.

**Reply by nwg-piotr** [reply]
Did it work for you on the first iso?

**Comment by Peppe61** [comment]
I am not sure. In the screen shot I made it is still Cantarell. But it is one of the things I usually change immediately ;-) I'll check and let you know tomorrow. No sound is a little inconvenient for now, so easy to check before I return to nwg on Arch ;-)

**Reply by nwg-piotr** [reply]
> changing the default font in GTK Settings (from Cantarell to any other font) doesn't seem to 'stick'; still Cantarell.

It works perfectly well for me on the install from the regular iso. Possibly it's related to the Sericea version only? I haven't yet tried it. @kuba3351, any idea?

**Reply by Peppe61** [reply]
This is, as it looks right now, a non-Fedora related thing. Changing the font in GTK settings does change it indeed (I used a crazy one so it was very visible). It doesn't change the font on the bottom bar and in the clock, but those are probably css defined. So I guess it is a feature request for the next nwg-shell. Sorry for the confusion, and I'm still on Sericea btw. I like it, works well!

**Reply by nwg-piotr** [reply]
> It doesn't change the font on the bottom bar and in the clock, but those are probably css defined. So I guess it is a feature request for the next nwg-shell.

I don't want panels to follow GTK theme settings, as they must stick out of the rest of GUI. Otherwise panels would look like windows. This must be defined in css, as we change more attributes than just the font face & size: colours, background & border attributes, whatever else css allows to define. If you want some panel element to follow GTK theme:

1. Either delete appropriate definition from the css style sheet, e.g.:

```css
#executor-label {
    font-family: DejaVu Sans Mono, monospace;
    font-size: 16px
}
```

2. Or give certain elements names, that don't exist in the css style sheet, e.g.:

![image](https://user-images.githubusercontent.com/20579136/217481610-85abbb96-ca55-40aa-a0d1-69622bab3856.png)

**Comment by kuba3351** [comment]
Hello

The difference of basic Fedora Sway Spin and the Sericea ostree variant is just a package manager, and in Sericea, the system partition is mounted read-only, only `/etc` and `/var` are writeable. The home partition is writeable because `/home` is a symlink to `/var/home`. So, if the settings are not stored outside of `/etc` and `/home` partition it should work similar.

I investigated the support of update indicator in Fedora, and I have a trouble. There is no easy way in `dnf` to count the number of packages that can be updated like iin `baph` The only command that is available is `dnf check-update`

Example output of this command is:
```
[root@NBJSIERZEGA jsierzega]# dnf check-update
Ostatnio sprawdzono ważność metadanych: 0:26:14 temu w dniu wto, 7 lut 2023, 21:55:36.

autocorr-en.noarch                                                                                 1:7.5.0.3-2.fc38                                                                                          rawhide
bash.x86_64                                                                                        5.2.15-3.fc38                                                                                             rawhide
dbus-broker.x86_64                                                                                 33-1.fc38                                                                                                 rawhide
device-mapper-persistent-data.x86_64                                                               0.9.0-10.fc38                                                                                             rawhide
fprintd.x86_64                                                                                     1.94.2-6.fc38                                                                                             rawhide
fprintd-pam.x86_64                                                                                 1.94.2-6.fc38                                                                                             rawhide
grub2-common.noarch                                                                                1:2.06-84.fc38                                                                                            rawhide
grub2-efi-ia32.x86_64                                                                              1:2.06-84.fc38                                                                                            rawhide
grub2-efi-x64.x86_64                                                                               1:2.06-84.fc38                                                                                            rawhide
grub2-pc.x86_64                                                                                    1:2.06-84.fc38                                                                                            rawhide
grub2-pc-modules.noarch                                                                            1:2.06-84.fc38                                                                                            rawhide
grub2-tools.x86_64                                                                                 1:2.06-84.fc38                                                                                            rawhide
grub2-tools-minimal.x86_64                                                                         1:2.06-84.fc38                                                                                            rawhide
ibus-typing-booster.noarch                                                                         2.21.0-1.fc38                                                                                             rawhide
iscsi-initiator-utils.x86_64                                                                       6.2.1.4-9.git2a8f9d8.fc38                                                                                 rawhide
iscsi-initiator-utils-iscsiuio.x86_64                                                              6.2.1.4-9.git2a8f9d8.fc38                                                                                 rawhide
kernel.x86_64                                                                                      6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kernel-core.x86_64                                                                                 6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kernel-devel.x86_64                                                                                6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kernel-modules.x86_64                                                                              6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kernel-modules-core.x86_64                                                                         6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kernel-modules-extra.x86_64                                                                        6.2.0-0.rc7.20230206gitd2d11f342b17.50.fc38                                                               rawhide
kf5-filesystem.x86_64                                                                              5.103.0-1.fc38                                                                                            rawhide
kf5-kwindowsystem.x86_64                                                                           5.103.0-1.fc38                                                                                            rawhide
libblkio.x86_64                                                                                    1.2.2-3.fc38                                                                                              rawhide
libreoffice.x86_64                                                                                 1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-base.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-calc.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-core.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-data.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-draw.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-emailmerge.x86_64                                                                      1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-graphicfilter.x86_64                                                                   1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-gtk3.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-gtk4.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-help-en.x86_64                                                                         1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-impress.x86_64                                                                         1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-langpack-en.x86_64                                                                     1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-math.x86_64                                                                            1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-ogltrans.x86_64                                                                        1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-opensymbol-fonts.noarch                                                                1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-pdfimport.x86_64                                                                       1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-pyuno.x86_64                                                                           1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-ure.x86_64                                                                             1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-ure-common.x86_64                                                                      1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-writer.x86_64                                                                          1:7.5.0.3-2.fc38                                                                                          rawhide
libreoffice-x11.x86_64                                                                             1:7.5.0.3-2.fc38                                                                                          rawhide
librsvg2.x86_64                                                                                    2.55.1-5.fc38                                                                                             rawhide
libserf.x86_64                                                                                     1.3.9-27.fc38                                                                                             rawhide
mdevctl.x86_64                                                                                     1.2.0-3.fc38                                                                                              rawhide
mercurial.x86_64                                                                                   6.3.2-3.fc38                                                                                              rawhide
poppler.x86_64                                                                                     23.02.0-1.fc38                                                                                            rawhide
poppler-cpp.x86_64                                                                                 23.02.0-1.fc38                                                                                            rawhide
poppler-glib.x86_64                                                                                23.02.0-1.fc38                                                                                            rawhide
poppler-utils.x86_64                                                                               23.02.0-1.fc38                                                                                            rawhide
redhat-rpm-config.noarch                                                                           247-1.fc38                                                                                                rawhide
rpm-ostree.x86_64                                                                                  2023.1-3.fc38                                                                                             rawhide
rpm-ostree-libs.x86_64                                                                             2023.1-3.fc38                                                                                             rawhide
rpm-sequoia.x86_64                                                                                 1.2.0-3.fc38                                                                                              rawhide
sord.x86_64                                                                                        0.16.14-1.fc38                                                                                            rawhide
thai-scalable-fonts-common.noarch                                                                  0.7.3-5.fc38                                                                                              rawhide
thai-scalable-waree-fonts.noarch                                                                   0.7.3-5.fc38                                                                                              rawhide
xorg-x11-server-Xorg.x86_64                                                                        1.20.14-18.fc38                                                                                           rawhide
xorg-x11-server-common.x86_64                                                                      1.20.14-18.fc38                                                                                           rawhide
zram-generator.x86_64                                                                              1.1.2-4.fc38                                                                                              rawhide
zram-generator-defaults.noarch                                                                     1.1.2-4.fc38                                                                                              rawhide
zvbi.x86_64                                                                                        0.2.35-19.fc38                                                                                            rawhide
Zastępowanie pakietów
grub2-tools-minimal.x86_64                                                                         1:2.06-84.fc38                                                                                            rawhide
    grub2-tools.x86_64                                                                             1:2.06-81.fc38                                                                                            @System
```

There is no pattern that I can use to `grep` and `wc -l` to get the number of packages especially the `dnf` is localised so any patterns based on the words will not work in other languages.

So, I think we should skip support of the update indicator right now and resolve remaining issues until someone will have an idea how to implement it properly.

**Reply by nwg-piotr** [reply]
No problem. If you don't install the `nwg-system-update` script, the related part of the form [just won't show up](https://github.com/nwg-piotr/nwg-shell-config/blob/41a4cb56fd22df2462bde231d79e6db9e9e4a767/nwg_shell_config/ui_components.py#L369).

![image](https://user-images.githubusercontent.com/20579136/217388579-158b5b25-bd2d-4a42-be38-e50f7fd5e9f5.png)

**Reply by nwg-piotr** [reply]
However, I need to set the `"update-indicator-on"` settings key `False`, as it's `True` by default now (my mistake!). It means we will need to release and use nwg-shell-config 0.4.15.

[edit] Done & released in nwg-shell-config 0.4.15.

**Comment by Peppe61** [comment]
I will try with sway-nwg (Rawhide) later today. Can be changed settings are overwritten from the immutable space on relogin.

**Reply by nwg-piotr** [reply]
Thanks. I should try the Sericea iso, but this night turned out too short. Maybe tomorrow.

**Comment by Peppe61** [comment]
Yes, exactly. Thanks for feedback. And I respect that pov, that is of course the uniqueness every dev of a spin needs to put into it. I already adjusted some css myself (for the bar color, to align it with the theme) and I will check the font. For now this 'question' can be closed. It is still a solid shell and a great wm to use.

**Comment by kuba3351** [comment]
I begin working to fix minor issues in Fedora. Created an bugzilla to fix azote package: https://bugzilla.redhat.com/show_bug.cgi?id=2169207

**Reply by nwg-piotr** [reply]
Very well. This is a minor issue, indeed. Any idea on how to add `pamixer`? I remember I couldn't install it due to some .so file missing.

**Reply by kuba3351** [reply]
Pamixer is available in another copr repo: https://copr.fedorainfracloud.org/coprs/opuk/pamixer/

We can add pamixer as dependency of nwg-shell package or enable this copr and explicitly install the pamixer in the installer script.

**Reply by nwg-piotr** [reply]
> We can add pamixer as dependency of nwg-shell package or enable this copr and explicitly install the pamixer in the installer script.

Yes, please. Otherwise the Controls module is broken on startup.

**Reply by kuba3351** [reply]
reated a PR for that: https://github.com/nwg-piotr/nwg-shell/pull/70

**Reply by nwg-piotr** [reply]
Merged. So, apart from the Azote missing icon issue, we're ready, right? I'll do a fresh install just in case, when I'm back home.

Are you by any chance active on Mastodon? There was some interest there, when I pre-announced the availability of nwg-shell on Fedora.

[fosstodon.org/@nwgpiotr](https://fosstodon.org/@nwgpiotr)

**Reply by kuba3351** [reply]
Yes, I can. But I have to create an account.

If you would like to do a clean install, I recommend to try with this ISO instead of linked in a wiki: https://dl.fedoraproject.org/pub/fedora/linux/development/38/Spins/x86_64/iso/

That's because the first Fedora version that will include Sway spin will be a Fedora 38. A few days go, the Fedora 38 was branched from rawhide, and the Rawhide is now a development version of Fedora 39. So, the Fedora 38 development now continues on a separate branch and this iso was build from that.

Also, keep in mind that during the development, until Fedora 38 will be officialy released, anything can break at any time.

**Reply by nwg-piotr** [reply]
Thank you. The Azote icon issue is actually not our matter. If my clean install goes well, I think we can consider nwg-shell for Fedora ready.

**Comment by kuba3351** [comment]
There is a commend about azote in bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=2169207 It is correct to have this file in described directory?

**Comment by nwg-piotr** [comment]
We want to reuse the icon in another GTK app, finding it by the icon name: 

```python
image = Gtk.Image.new_from_icon_name(name, size)
```

For this to work, we need an icon of this name to be a part of the GTK icon theme (impossible), or to be placed in `/usr/share/pixmaps/`. I don't know about any other solution.

**Reply by kuba3351** [reply]
Ok, can you create an account in the bugzilla and write a comment? I think it is not necessary to copy and paste content from here to the bugzilla, let's talk directly with the package author to resolve the issue :)

**Comment by nwg-piotr** [comment]
Sure. But before I do so, could you please paste the content of `/usr/share/applications/azote.desktop` here? Mean the Fedora version of course. I have no Linux machine in range at the moment. :(

**Reply by kuba3351** [reply]
Here it is:

```
[jsierzega@NBJSIERZEGA ~]$ cat /usr/share/applications/azote.desktop
[Desktop Entry]
Encoding=UTF-8
Type=Application
Exec=azote
Name=Azote
Icon=/usr/share/azote/azote.svg
Terminal=false
GenericName=Azote wallpaper manager
GenericName[pl]=Menedżer tapet Azote
GenericName[de]=Azote Bildschirmhintergrund-Manager
Categories=Utility;GTK;DesktopSettings;
Comment=Browse, flip and set desktop wallpapers
Comment[pl]=Przeglądaj, odwracaj i ustawiaj tapety pulpitu
Keywords=background;desktop;
X-Desktop-File-Install-Version=0.26
```

**Comment by nwg-piotr** [comment]
Thanks! Bob replaced:

```
Icon=azote
```
with:

```
Icon=/usr/share/azote/azote.svg
```

That's why we have the launcher icon. But the icon name is unavailable. I'm leaving the office now, but will sign up on bugzilla and clarify what the problem is, as soon, as I can.

**Reply by kuba3351** [reply]
Thank you

**Comment by nwg-piotr** [comment]
It installed perfectly well.

Are we going to have [nwg-shell-config 0.4.15](https://github.com/nwg-piotr/nwg-shell-config/releases/tag/v0.4.15) any time soon? If so, could you check if the `/usr/share/pixmaps nwg-shell-translate.svg` icon is being installed?

**Comment by t0fik** [comment]
I did not know, that my packages make such impact :). Today I've setup automation for building packages, so packages in my copr `tofik/nwg-shell` should be in sync with project repositories with minimal delay (up to 1 day), if building process do not change, eg. changes in dependencies.
To `nwg-shell` package, I've added patch removing dependency on `pamixer` on Fedora.

**Comment by nwg-piotr** [comment]
I'm not sure if `pamixer` is needed on Fedora. @kuba3351, do you agree?

By the way: I'm adding Hyprland support starting from nwg-shell 0.5.0. What's Hyprland status on Fedora? We need v0.26.0 for all the stuff to work properly.

**Reply by t0fik** [reply]
Just to add some details. I'm using `pactl set-sink-volume 0 [+|-]5%` on Fedora, `pactl` is part of `pulseaudio-utils` package.
IMHO there is no need for `pamixer` dependency.

Hyprland do not have official packages in Fedora, I've found review package on COPR (https://copr.fedorainfracloud.org/coprs/g/fedora-review/fedora-review-2192986-hyprland/), but it's not building and some user COPRs. In one word, doesn't look good.

**Comment by nwg-piotr** [comment]
If I remember well, `pamixer` is a panel dependency, not necessarily the shell itself.

**Reply by t0fik** [reply]
Yes, `nwg-panel` requires `pamixer`. I should have some time this week and I can rewrite this to utilize `pactl`, which is part of `libpulse` on Arch, one of dependecies of `pamixer` and `nwg-panel` is already calling `pactl` in other part of the code.

**Reply by t0fik** [reply]
Patched Fedora package is in my COPR (`tofik/nwg-shell`). Also PR adding support for `pactl` to interact with audio is on place https://github.com/nwg-piotr/nwg-panel/pull/229

**Reply by nwg-piotr** [reply]
@t0fik Yeah, sorry for making you wait with your PRs. A lot for work and I'm short of time lately. Will check and merge ASAP.

**Reply by t0fik** [reply]
@nwg-piotr No problem, we all doing that in our free time.

**Reply by nwg-piotr** [reply]
> Patched Fedora package is in my COPR (`tofik/nwg-shell`).

If you think it should be made public, feel free to edit Wiki.

**Reply by t0fik** [reply]
I've updated wiki pages and created this issue in `nwg-panel`: https://github.com/nwg-piotr/nwg-panel/issues/230

Patching code during building packages is not needed any more, due to that all my PRs adding support for `pactl` or removing dependency for `pamixer` are merged already in `nwg-shell` since 0.5.0 and in `nwg-panel` since 0.9.7
